compiler/src/dmd/timetrace.d

```diff
@@ -0,0 +1,482 @@
+/**
+Compilation time tracing, --ftime-trace.
+
+The time trace profile is output in the Chrome Trace Event Format, described
+here: https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview
+
+This file is originally from LDC (the LLVM D compiler).
+
+Copyright: Copyright (C) 1999-2022 by The D Language Foundation, All Rights Reserved
+Authors:   Johan Engelen, Max Haughton
+License:   $(LINK2 https://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+Source:    $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/common/timetrace.d, common/_timetrace.d)
+Documentation: https://dlang.org/phobos/dmd_common_timetrace.html
+Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/common/timetrace.d
+*/
+module dmd.timetrace;
+
+import dmd.errors;
+import dmd.globals;
+import dmd.location;
+import dmd.root.array;
+import dmd.root.file;
+import dmd.common.outbuffer;
+import dmd.root.string : toDString;
+
+// Thread local profiler instance (multithread currently not supported because compiler is single-threaded)
+TimeTraceProfiler* timeTraceProfiler = null;
+
+// processName pointer is captured
+extern(C++)
+void initializeTimeTrace(uint timeGranularity, uint memoryGranularity, const(char)* processName)
+{
+    assert(timeTraceProfiler is null, "Double initialization of timeTraceProfiler");
+    timeTraceProfiler = new TimeTraceProfiler(timeGranularity, memoryGranularity, processName);
+}
+
+extern(C++)
+void deinitializeTimeTrace()
+{
+    if (timeTraceProfilerEnabled())
+    {
+        object.destroy(timeTraceProfiler);
+        timeTraceProfiler = null;
+    }
+}
+
+pragma(inline, true)
+extern(C++)
+bool timeTraceProfilerEnabled()
+{
+    version (LDC)
+    {
+        import ldc.intrinsics: llvm_expect;
+        return llvm_expect(timeTraceProfiler !is null, false);
+    }
+    else
+    {
+        return timeTraceProfiler !is null;
+    }
+}
+
+const(char)[] getTimeTraceProfileFilename(const(char)* filename_cstr)
+{
+    const(char)[] filename;
+    if (filename_cstr)
+    {
+        filename = filename_cstr.toDString();
+    }
+    if (filename.length == 0)
+    {
+        if (global.params.objfiles.length)
+        {
+            filename = global.params.objfiles[0].toDString() ~ ".time-trace";
+        }
+        else
+        {
+            filename = "out.time-trace";
+        }
+    }
+    return filename;
+}
+
+extern(C++)
+void writeTimeTraceProfile(const(char)* filename_cstr)
+{
+    if (!timeTraceProfiler)
+        return;
+
+    const filename = getTimeTraceProfileFilename(filename_cstr);
+
+    OutBuffer buf;
```

Comment:
Consider returning an OutBuffer, or moving this `buf` local to a `ref` parameter.  Doing IO inside public API is regarded as bad style, and rules it out from being unittest-able.

Reply:
good point

Reply:
I moved file IO to main.d